translate/storage/stringsdict.py

```diff
@@ -0,0 +1,154 @@
+    number of variables with plural strings.
+
+    Each entry is split up into multiple translation units, containing either
+    the format string or one of the variables.
+    """
+
+    UnitClass = StringsDictUnit
+    Name = "iOS Stringsdict"
+    Mimetypes = ["application/x-plist"]
+    Extensions = ["stringsdict"]
+
+    def __init__(self, inputfile=None, **kwargs):
+        super().__init__(**kwargs)
+        self.parse(inputfile)
+
+    def gettargetlanguage(self):
+        target_lang = super().gettargetlanguage()
+
+        # If targetlanguage isn't set, we try to extract it from the filename path (if any).
+        if target_lang is None and hasattr(self, "filename") and self.filename:
+            parent_dir = os.path.split(os.path.dirname(self.filename))[1]
+            match = re.search(r"^(\w*).lproj", parent_dir)
+            if match is not None:
+                target_lang = match.group(1)
+            else:
+                target_lang = self.sourcelanguage
+
+            # Cache it
+            self.settargetlanguage(target_lang)
+
+        return target_lang
+
+    @property
+    def target_plural_tags(self):
+        """Get all supported plural tags for the target language.
+        Note that 'zero' is always supported.
+        """
+        target_lang = self.gettargetlanguage()
+        if target_lang is None:
+            return data.cldr_plural_categories
+
+        locale = target_lang.replace("_", "-").split("-")[0]
+        tags = data.plural_tags.get(locale, data.cldr_plural_categories)
+        if "zero" not in tags:
+            tags.insert(0, "zero")
+        return tags
+
+    def parse(self, input):
+        """Read a .stringsdict file into a dictionary, and convert it to translation units."""
+
+        if isinstance(input, bytes) or isinstance(input, str):
+            plist = plistlib.loads(input, dict_type=OrderedDict)
+        elif input is not None:
+            plist = plistlib.load(input, dict_type=OrderedDict)
+        else:
+            plist = {}
+
+        for key, outer in plist.items():
+            if not isinstance(outer, dict):
+                raise ValueError(f"{key} is not a dict")
+            for innerkey, value in outer.items():
+                if innerkey == "NSStringLocalizedFormatKey":
+                    u = self.UnitClass(key)
+                    u.target = str(value)
+                    self.addunit(u)
+                elif isinstance(value, dict):
+                    spec_type = value.get("NSStringFormatSpecTypeKey", "")
+                    if spec_type and spec_type != "NSStringPluralRuleType":
+                        raise ValueError(
+                            f"{innerkey} in {key} is not of NSStringPluralRuleType"
+                        )
+
+                    plural_tags = self.target_plural_tags
+                    plural_strings = [value.get(tag, "") for tag in plural_tags]
+
+                    u = self.UnitClass(f"{key}[{innerkey}]")
```

Comment:
Can the key also contain [] characters? In case it can, it can break easily. DictUnit / DictStore classes handle this better.

Reply:
I've changed the base class to DictUnit, and added a UnitId subclass using `:` as a separator (in line with the `plistbuddy` command line tool). The parser now sets the unitid directly.